### PR TITLE
Add review date widget and needs review view.

### LINKED ISSUE #3: Workflow Content lifecycle: needs review status and dashboard
### User story

**As a:** Product Owner,
**I want:** A review date and needs review status for content to be enabled by default when I enable localgov_workflows
**So that:** there is a mechanism to keep content up to date and quality assured

### Notes

Content lifecycle refers to functionality to ensure that content is always up to date and relevant to end users.

### Acceptance tests

We want a way for an Editor [editor] to ensure that content is regularly reviewed, updated on a timely basis, and removed when no longer useful to the end user.

To achieve this, there must be a field for all pages called review date. By default, a council can choose a specific policy for all content to be reviewed. This period could be 12 months, for example. In this instance, every page that is created or updated will automatically have the review date applied to it.

A Contributor [contributor], Author [author], or Editor [editor] must also have the option to choose a different review date to their liking, whenever they edit. They can choose a specific date or timeframe eg. 3 months | 6 months | 1 year | 18 months | 2 years | 3 years

When the review date for content passes, it automatically has a status applied to it of “needs review”.

An Editor [editor] must be able to see a view that shows all pages that need reviewing. From that view, they can decide what to do with content in review.

### Tasks

- [x] Create configuration (programatically) for **2 date fields** across all content types with workflow enabled. Two fields: last review date, next review date, plus a check box to indicate that this content has been reviewed.
- [x] Configure global default review date (for example 12 months) with option to change to other timeframes, probably on an admin form
- [x] Ensure permissions allow appropriate roles to set review dates
- [x] Create 'needs review' dashboard and set permissions to view it. Possibly cleanest to have a separate list for content with a review date in the past. 
- [x] Write update hooks to add new config to existing installs.
- [x] Add some custom form_alter (or equivalent) to process the checkbox and set date fields accordingly. 
- [x] Write cron job to update content that has past the sell by date and set to state to 'needs review'.
- [x] Write tests.
- [x] Confirm that a Contributor [contributor], Author [author], or Editor [editor] can choose a different review date to their liking, whenever they edit.

### PR DESCRIPTION
**Pull request by finnlewis**
Adds a review date widget which allows users to mark content as reviewed and specify a date when the content next needs to be reviewed. Also includes a needs review view (/admin/content/localgov_review) and an admin page to set the default review date (/admin/config/workflow/localgov-review-date).

When marking content as reviewed and setting the next review date, a scheduled transition is created that will move the content to 'needs review' on the review date.

In the backend, it creates a ReviewDate entity that records when an entity was reviewed, when it is next due for review and the scheduled transition associated with the review.

### CODE REVIEW

**Comment by stephen-cox** [comment]
@finnlewis This is ready for review now. It doesn't have complete test coverage, but there are tests for major stuff. Could do with JavaScript tests for testing the review date widget.

**Comment by stephen-cox** [comment]
Hmm ... tests failing because this patch is failing to apply: https://github.com/localgovdrupal/localgov_workflows/blob/1c41530396e9bab26c9ce36d148d1936c6d47e97/composer.json#L23

Will investigate when I get a chance

**Comment by finnlewis** [comment]
I think patches for core need to be in the composer.json that requires core, the profile in our case:
 
https://github.com/localgovdrupal/localgov/blob/2.x/composer.json#L54

**Comment by stephen-cox** [comment]
> I think patches for core need to be in the composer.json that requires core, the profile in our case:

Yep - moved core patch to https://github.com/localgovdrupal/localgov/pull/373 which will need to merged for tests to pass.

**Comment by finnlewis** [comment]
This is ready to be merged, but @ekes would like to take a little time to review the code. 

It would also be really good to test against an existing install: 
@Adnan-cds @andybroomfield @j4-m @davidsiddall any of you able to review also?

**Comment by Adnan-cds** [comment]
Testing...

**Comment by finnlewis** [comment]
We are aiming to merge this and release next Monday, to bring it in sprint.

Please could you review if you have time before then?

**Comment by Adnan-cds** [comment]
I have finished testing the functionality.  Works well when I update publishing states from the node edit form.  Nothing happens when I do this from the "Latest revision" tab.  I am [attaching a screenshot](https://user-images.githubusercontent.com/50206849/135256798-678d4f8b-bc8e-4480-a10c-2cb634d54b4e.png).

There is no explicit way to say that a page does *not* need any further review.  I can uncheck the "Content reviewed" checkbox on the node edit form's sidebar, but that doesn't exactly convey the intention and can potentially cause confusion to editors.  I could be wrong on this assumption.  I will leave this to whoever drafted the user story.

I haven't reviewed the code in detail.  It looked okay yesterday at first glance.  I will leave a detailed review for Ekes.

**Comment by ekes** [comment]
> I have finished testing the functionality. Works well when I update publishing states from the node edit form. Nothing happens when I do this from the "Latest revision" tab. I am attaching a screenshot.

I'm guessing that would require using the default date, and doing so in the save events? Maybe a feature request as this presently works just with the form? Or could it even be a default value logic? Something like that.

File: modules/localgov_review_date/localgov_review_date.module
Comment (ekes) [inline modules/localgov_review_date/localgov_review_date.module]: Seems that Scheduled Transitions doesn't quite work like this. It will store configuration for content type bundles that don't have enabled transitions.

Looking at how Scheduled Transitions does it on their config form https://git.drupalcode.org/project/scheduled_transitions/-/blob/1b3d8c51108ad7dbf87611e1fac14410592d9c2f/src/Form/ScheduledTransitionsSettingsForm.php#L140 they're using a utility service `scheduled_transitions.utility`

File: modules/localgov_review_date/src/Entity/ReviewDate.php
Comment (ekes) [inline modules/localgov_review_date/src/Entity/ReviewDate.php:70]: Being picky, but this and later, EntityBase already has the entity type manager.

File: modules/localgov_review_date/src/Entity/ReviewDate.php
Comment (ekes) [inline modules/localgov_review_date/src/Entity/ReviewDate.php]: Being picky but the EntityBase already has the entity type manager.

File: modules/localgov_review_date/tests/src/FunctionalJavascript/ReviewDateWidgetTest.php
Comment (ekes) [inline modules/localgov_review_date/tests/src/FunctionalJavascript/ReviewDateWidgetTest.php]: I fear this test, probably the test rather than the code it's testing, might be prone to underlying timezone failures. 

I'm presently getting:

--- Expected
+++ Actual
@@ @@
-'2022-01-02'
+'2022-01-01'

I'll investigate further. IIRC Drupal tests, or at least core, was running with a timezone of Australia/Sydney

Comment (ekes) [reply modules/localgov_review_date/tests/src/FunctionalJavascript/ReviewDateWidgetTest.php]: Yep code isn't happy in the early Australian morning https://git.drupalcode.org/project/drupal/-/blob/038cb48a4a640fd4913e0b6284f60f6fe82fc9cf/core/tests/bootstrap.php#L175 I'd say. Here github run at round 03:00 is Sydney
https://github.com/localgovdrupal/localgov_workflows/pull/18/checks?check_run_id=3769485313

File: modules/localgov_review_date/src/Plugin/Field/FieldWidget/ReviewDateWidget.php
Comment (ekes) [inline modules/localgov_review_date/src/Plugin/Field/FieldWidget/ReviewDateWidget.php:110]: Rather than patching core https://github.com/localgovdrupal/localgov/pull/373 which also only works if you have installed the profile, can we not just add
`  'type' => 'date',`
here into the '#attributes'?

**Comment by stephen-cox** [comment]
>   I have finished testing the functionality. Works well when I update publishing states from the node edit form. Nothing happens when I do this from the "Latest revision" tab. I am attaching a screenshot.

> I'm guessing that would require using the default date, and doing so in the save events? Maybe a feature request as this presently works just with the form? Or could it even be a default value logic? Something like that.

This was discussed briefly by the working group. It was decide that checking the content reviewed box when publishing the content was more of a UX shortcut for now rather than enforcing content is reviewed whenever it's published. I suggest this can be moved to a new issue for enforcing content review when publishing content and/or the UX of this needs further work.

File: modules/localgov_review_date/localgov_review_date.module
Comment (stephen-cox) [reply modules/localgov_review_date/localgov_review_date.module]: I have update this to use `$scheduled_bundles = \Drupal::service('scheduled_transitions.utility')->getBundles();`

File: modules/localgov_review_date/src/Entity/ReviewDate.php
Comment (stephen-cox) [reply modules/localgov_review_date/src/Entity/ReviewDate.php]: Yep - updated

File: modules/localgov_review_date/tests/src/FunctionalJavascript/ReviewDateWidgetTest.php
Comment (stephen-cox) [reply modules/localgov_review_date/tests/src/FunctionalJavascript/ReviewDateWidgetTest.php]: Updated to use a set timestamp rather than now

File: modules/localgov_review_date/src/Plugin/Field/FieldWidget/ReviewDateWidget.php
Comment (stephen-cox) [reply modules/localgov_review_date/src/Plugin/Field/FieldWidget/ReviewDateWidget.php:110]: That seems to work.

File: modules/localgov_review_date/tests/src/FunctionalJavascript/ReviewDateWidgetTest.php
Comment (stephen-cox) [reply modules/localgov_review_date/tests/src/FunctionalJavascript/ReviewDateWidgetTest.php]: I have set the timezone in Drupal to try and fix this issue. The tests pass now, but I don't know if this us just because they didn't hit the timezone issue. Not sure what else can be done here beyond removing this test.

https://github.com/localgovdrupal/localgov_workflows/blob/e683fae983afa696cf3d08055432bde4bb529d5e/modules/localgov_review_date/tests/src/FunctionalJavascript/ReviewDateWidgetTest.php#L73-L75

**Comment by Adnan-cds** [comment]
> I suggest this can be moved to a new issue for enforcing content review when publishing content and/or the UX of this needs further work.

Fine with me.

File: modules/localgov_review_date/tests/src/FunctionalJavascript/ReviewDateWidgetTest.php
Comment (stephen-cox) [reply modules/localgov_review_date/tests/src/FunctionalJavascript/ReviewDateWidgetTest.php]: I re-ran the tests and they failed with the timezone issue: https://github.com/localgovdrupal/localgov_workflows/pull/18/checks?check_run_id=3817280968

I'm going to comment out the test so we can get this reviewed and merged, but if someone has any idea how to solve it please dive in.

**Comment by finnlewis** [comment]
This is all working for me. The tests pass. @Adnan-cds and @ekes your comments suggest that your happy with this, can I get an official approval and I'll merge it? (I created the original merge request, so I'll see it through) 

Thanks dudes!

**Comment by Adnan-cds** [comment]
Feel free to merge once @ekes approves.  Saying this because Ekes has done a details review unlike me.

**Comment by andybroomfield** [comment]
I may be being silly, but how do I see the review date widgets. I've turned on the module, but is there some config per content type to turn on or should I see it already?

**Comment by Adnan-cds** [comment]
> but how do I see the review date widgets. 

It's on the sidebar of the node edit form.  I had the same issue and was about to give up when I noticed it :)

**Comment by andybroomfield** [comment]
I have LocalGov Review Date switched on and this is a node edit screen. Do I need to switch something on?

screencapture-localgov-lndo-site-8000-node-96-edit-2021-10-08-15_55_13

**Comment by Adnan-cds** [comment]
That doesn't look right Andy.  Could you please check the Form display for the Service page content type?  The Review date field appears at the top of the list.  I am [attaching](https://user-images.githubusercontent.com/50206849/136580571-3202fe5f-7517-4897-a95b-26cc7ca127ff.png) what I am seeing.

**Comment by andybroomfield** [comment]
@Adnan-cds No review date field. Confirm the module is switched on (just drush pmu localgov_review_date and drush en localgov_review_date).
So is there a new field type I need to add to each content type to get this functionality?

**Comment by Adnan-cds** [comment]
Also, please check the default review date from /admin/config/workflow/localgov-review-date.  I can't remember if this is a must do before the thing works.

**Comment by andybroomfield** [comment]
@Adnan-cds That just has a setting for the default to 1 year, I've clicked save and not seeing any change.

**Comment by Adnan-cds** [comment]
I may have spotted the issue from the screenshot you have attached.  There's no "Scheduled transition" tab in the node edit form.  This Review date functionality relies on scheduled transition.  Any chance you can activate Scheduled transition on a content type and retest.

**Comment by andybroomfield** [comment]
@Adnan-cds Ahh, the module scheduled transitions was installed, but not enabled for any node types. 
This is on a fresh install (lando drush si localgov, lando drush en localgov_demo, lando drush en localgov_workflows, lando drush en localgov_review_date) so I think that needs to be either documented in the readme, or if workflows is switched on add all the localgov content types to scheduled transitions.

**Comment by andybroomfield** [comment]
Ok, so after getting scheduled transitions eneabled for each content type I see the review widget.
I was able to trick it into review by setting a date in the past and then it goes into review.
One thing that I think is odd, when I select I have reviewed this content, but don't select publish in workflow, the page is not republished, which is what I would expect to happen.

File: modules/localgov_review_date/src/Plugin/Field/FieldType/ReviewDateItem.php
Comment (andybroomfield) [inline modules/localgov_review_date/src/Plugin/Field/FieldType/ReviewDateItem.php:113]: Is there a Drupal version of strtotime?
We should make sure we get the the review date based on the sites timezone, either here or in the entity.

Comment (ekes) [reply modules/localgov_review_date/src/Plugin/Field/FieldType/ReviewDateItem.php:113]: I think normally it's to put it into a https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Datetime%21DrupalDateTime.php/class/DrupalDateTime object.

Comment (ekes) [reply modules/localgov_review_date/src/Plugin/Field/FieldType/ReviewDateItem.php:113]: So the timezone logic here if it is to work outside of UTC / UTC+1 without noticeable quirks:

 * The review date is stored as a unix timestamp - as it's a scheduled transition, although the review entity also uses timestamp.
 * The timestamp is calculated from (the 1 second out and it's a different day) 00:00 which is when the transition is expected to be triggered.

 * The timestamp should be in the Site Timezone to trigger the transition at 00:00 Site Time.
 * When the timestamp converted to YYYY-MM-DD on the form it should thus use the Site Timezone.
 * When the form is saved YYYY-MM-DD should be converted to unix timestamp based on the Site TImezone.

At the moment Drupal runs date_default_timezone_set() with in order of preference: user timezone, site timezone, system timezone. So consistently using strtotime() - including in the test - with a site timezone set, and no user timezone, should consistently return the expected value. 

As soon as a user timezone is thrown into the mix the bets are off; and this is also then the javascript calculation https://github.com/localgovdrupal/localgov_workflows/blob/f1e4f7331fea34e7b3a9b8fdd4d851c09948f337/modules/localgov_review_date/js/review-date.js#L38 of time in the 'testing browser'?

**Comment by finnlewis** [comment]
Just chatting with @ekes at standup, we're hoping to merge this today at 'merge monday' and open issues for the edge cases discussed above.

File: modules/localgov_review_date/src/Plugin/Field/FieldType/ReviewDateItem.php
Comment (ekes) [reply modules/localgov_review_date/src/Plugin/Field/FieldType/ReviewDateItem.php:113]: I've opened this as an issue to be fixed - if need be later.

**Review by ekes — APPROVED** [review]
With the new issue about timezones all other issues raised have now been addressed.

**Pull request merged**
